vernac/lemmas.ml

```diff
@@ -299,20 +299,14 @@ let save_remaining_recthms env sigma ~poly ~scope norm univs body opaq i
         let kn = declare_constant name ~local (DefinitionEntry const, k) in
         (ConstRef kn,impargs)
 
-let initialize_named_context_for_proof () =
-  let sign = Global.named_context () in
-  List.fold_right
-    (fun d signv ->
-      let id = NamedDecl.get_id d in
-      let d = if variable_opacity id then NamedDecl.drop_body d else d in
```

Comment:
This is to make the following fail:
~~~coq
Section foo.

  Let bla : nat.
  Proof. exact 0. Qed.

  Lemma bli : bla = 0.
  Proof. reflexivity. (* unable to unify 0 with bla *) Qed.
End foo.
~~~
Of course this is kinda broken when code calls start_lemma with an explicit `sign`, for instance the following succeeds in master:
~~~coq
Require Import Program.Tactics.
Class MyEq {A} a b := myeq : a = b :> A.

Section foo.

  Let bla : nat.
  Proof. exact 0. Qed.

  Program Instance bli : MyEq bla 0.
  Check bli.
End foo.
~~~

Maybe we should just deprecate Qed for Let?

Reply:
Actually the code in shape so that we could make `Definition` and `Program` respect the setting; but indeed, I am not sure I do understand the purpose of `Let` in particular the purpose of interactive "Let" when we already have `Lemma` and `Definition`. So maybe we should just deprecate the whole `Let` construction.

Reply:
>in particular the purpose of interactive "Let" when we already have Lemma and Definition.

Let produces section letins (in sections). The issue is that Qed is broken for these, especially as after the section is closed they become regular letins regardless of Qed.

Reply:
Yeah so how useful are these let ins, at least in the developments I know their usage seems a bit marginal even if authors do likely appreciate their use. But indeed they cannot be inlined at the top level, so wouldn't a (private) definition be better?

Reply:
Transparent letins are fine though